### ISSUE TITLE
The reproducibility of experiments in the paper

### ISSUE THREAD

**Issue by neiljohn1990** [bboylyg/NAD#11]
Hi author, I have some questions about this paper and the public code.

Your v1 version of arXiv paper is **15 Jan 2021** ([https://arxiv.org/abs/2101.05930v1](https://arxiv.org/abs/2101.05930v1)), and your first commit on the Github is **21 Jan 2021** ([https://github.com/bboylyg/NAD/commit/a5a10f1ee746d51e75945821027ade76e88e2901](https://github.com/bboylyg/NAD/commit/a5a10f1ee746d51e75945821027ade76e88e2901)).

The most important loss that you proposed in your paper is Eq(3), which is controlled by the hyperparameter beta. 

However, in your first commit code, this part is written as :
```python
cls_loss = criterionCls(output_s, target)  
at3_loss = criterionAT(activation3_s, activation3_t).detach() * opt.beta3  
at2_loss = criterionAT(activation2_s, activation2_t).detach() * opt.beta2  
at1_loss = criterionAT(activation1_s, activation1_t).detach() * opt.beta1  
at_loss = at1_loss + at2_loss + at3_loss + cls_loss
```
The `detach()` in Pytorch is used to return a new Tensor, detached from the current graph [(see the doc).](https://pytorch.org/docs/stable/generated/torch.Tensor.detach.html). So if the model uses this `at_loss` to optimize, the `at1_loss, at2_loss, at3_loss` will not contribute anything to the training of the model. The user @zeabin has submit the issue [https://github.com/bboylyg/NAD/issues/8](https://github.com/bboylyg/NAD/issues/8) and fortunately, you fixed it in [https://github.com/bboylyg/NAD/commit/6907ea2fb9c8445fb408fd77627e4f990ad6d9be](https://github.com/bboylyg/NAD/commit/6907ea2fb9c8445fb408fd77627e4f990ad6d9be) at **10 Jan 2022**:
```python
at3_loss = criterionAT(activation3_s, activation3_t.detach()) * opt.beta3
at2_loss = criterionAT(activation2_s, activation2_t.detach()) * opt.beta2
at1_loss = criterionAT(activation1_s, activation1_t.detach()) * opt.beta1
```

Based on the above facts, my question is **whether the results in your paper are based on the first commit code or the fixed code**.  

1. If the experiments were run with the correct code, why the first commit is wrong?

2. If the experiments were run with the wrong code, the idea does not work in your paper, it's just a fine-tuning. I have valid reasons to doubt the reliability of the results.

3. For the papers published from 21 Jan 2021 -- 10 Jan 2022, and use NAD as a comparison, whether the results of these papers are reliable. Because during this time, the code in this repository is wrong.

I will appreciate it if you can solve my above questions.

**Comment by bboylyg** [comment]
Hi,
Thank you for paying attention to our work. The response to your questions are as follows:
1.  We need to make it clear that the error in the first commit is only a **code writing error**, and we appreciate you pointing it out. In fact, the initial NAD code is modified by the open-sourced [knowledge distillation library ](https://github.com/AberHu/Knowledge-Distillation-Zoo) which is used a correct loss formulation you can check.
2. Our experiments are correctly based on the right loss formulation and we hope you can run the experiments directly to verify the NAD's performance.
3. We sincerely apologize for this careless act and will identify the implications for other research.

**Comment by TakaSaitama** [comment]
Hello, could you kindly upload a license file ? an open one would be appreaciated.
